### DISCUSSION TITLE
How to ensure the coordinate origin of a PDF page remains stable?

### DISCUSSION THREAD

**Discussion by TCHSDUFH** [pymupdf/PyMuPDF#3937] [Looking for help]
Hi, developers! I encountered an issue while using PyMuPDF. Some pages in my PDF document have been rotated or flipped upside down. When I try to add page numbers to the bottom-left corner, the position calculation still uses the coordinate origin from before the page was rotated, which causes the page numbers to be placed in the wrong position. How can I ensure that the coordinate origin is always at the top-left corner, or are there other ways to make sure the page number is inserted at a stable position? My demo code⬇
```Python
page_width = page.rect.width
page_height = page.rect.height

rect = fitz.Rect(sideGap, page_height - buttonGap - height, sideGap + width, page_height - buttonGap)

page.draw_rect(rect, color=(1, 1, 1), fill=(1, 1, 1))
page.insert_textbox(rect,
                            nowPage,
                            fontsize=int(pageFontSet.fontSize),
                            fontname=finalFont,   
                            color=pageFontSet.fontColor,   
                            align=fitz.TEXT_ALIGN_CENTER,  
                            overlay=True)  
```
![093212](https://github.com/user-attachments/assets/ae287b4b-fa7c-4627-814f-aad7158e1558)

**Comment by JorjMcKie** [comment]
You can find out whether a page is rotated (`page.rotation`). You can also de-rotate a page without changing appearance, and then inert your page number like in the normal unrotated case.
So far so easy.

I have no idea how to determine a left-right or top-bottom flip however.
But `page.insert_textbox` and the `page.draw_*` allow you to "morph" their output via `morph=(point, matrix)`. If you choose `point = pymupdf.Point(0, page.rect.height/2)` and `matrix = pymupdf.Matrix(1, -1)` (top-bottom flip), then you should achieve this.

**Comment by ffalcon31415** [comment]
I am struggling with the same issue. 
I tried to apply page.remove_rotation() before page.insert_textbox(), but it still is putting the textbox rotated on pages that have page.rotation != 0

I would have thought that was the solution but clearly I'm missing something.

After looking through the docs I do not understand how to use the morph=(point,matrix) parameter ... is there somewhere to look for examples of that or documentation of the morph parameter?

Thanks to all contributors for this excellent tool.

----

*Edit*

Apologies -- page.remove_rotation() did in fact solve my issue (my mistake in believing it did not)

**Reply by TCHSDUFH** [reply]
Do you have a solution? For now, I can only save the PDF as PDF/A-2 format through other software before adding the page numbers where I want them.

**Reply by ffalcon31415** [reply]
for me, page.remove_rotation() prior to page.insert_textbox() did fix the issue I was having. Sorry
